rfcs/0000-custom-health-checks/README.md

```diff
@@ -0,0 +1,298 @@
+
+```yaml
+---
+apiVersion: cert-manager.io/v1
+kind: Certificate
+metadata:
+  name: app-certificate
+spec:
+  commonName: cert-manager-tls
+  dnsNames:
+  - app.ns.svc.cluster.local
+  ipAddresses:
+  - x.x.x.x
+  isCA: true
+  issuerRef:
+    group: cert-manager.io
+    kind: ClusterIssuer
+    name: app-issuer
+  secretName: app-tls-certs
+  subject:
+    organizations:
+    - example.com
+```
+
+This `Certificate` resource will transition through the following `conditions`: 
+`Issuing` and `Ready`.
+
+In order to compute the status of this resource, we need to look at both the `Issuing`
+and `Ready` conditions.
+
+The Flux `Kustomization` object used to apply the `Certificate` will look like this:
+
+```yaml
+apiVersion: kustomize.toolkit.fluxcd.io/v1
+kind: Kustomization
+metadata:
+  name: certs
+spec:
+  interval: 5m
+  prune: true
+  sourceRef:
+    kind: GitRepository
+    name: flux-system
+  path: ./certs
+  wait: true
+  healthCheckExprs:
+  - apiVersion: cert-manager.io/v1
+    kind: Certificate
+    inProgress: "status.conditions.filter(e, e.type == 'Issuing').all(e, e.observedGeneration == metadata.generation && e.status == 'True')"
+    failed: "status.conditions.filter(e, e.type == 'Ready').all(e, e.observedGeneration == metadata.generation && e.status == 'False')"
+    current: "status.conditions.filter(e, e.type == 'Ready').all(e, e.observedGeneration == metadata.generation && e.status == 'True')"
+```
+
+The `.spec.healthCheckExprs` field contains an entry for the `Certificate` kind, its `apiVersion`,
+and the CEL expressions that need to be met in order to determine the health status of all custom resources
+of this kind reconciled by the Flux `Kustomization`.
+
+Note that all the Kubernetes core resources are discarded from the `healthCheckExprs` list.
+
+### Custom Health Check Library
+
+To help users define custom health checks, we will provide on the [fluxcd.io](https://fluxcd.io)
+website a library of custom health checks for popular custom resources.
+
+The Flux community will be able to contribute to this library by submitting pull requests
+to the `fluxcd/website` repository. 
+
+### User Stories
+
+#### Configure health checks for non-standard custom resources
+
+> As a Flux user, I want to be able to specify health checks for
+> custom resources that don't have a Ready condition, so that I can be notified
+> when the status of my resources transitions to a failed state based on the evaluation
+> of a different condition.
+
+Using `.spec.healthCheckExprs`, Flux users have the ability to
+specify the conditions that need to be met in order to determine the status of
+a custom resource. This enables Flux to query any `.status` field,
+besides the standard `Ready` condition, and evaluate it using a CEL expression.
+
+Example for `SealedSecret` which has a `Synced` condition:
+
+```yaml
+  - apiVersion: bitnami.com/v1alpha1
+    kind: SealedSecret
+    inProgress: "metadata.generation != status.observedGeneration"
```

Comment:
note that checking the generation for `inProgress` is done by default by in the POC. But as we perform the evaluation with the whole object, any field can used in the expression.

Reply:
Not all CRDs have `status.observedGeneration` for example one of Flux Operator nor ClusterAPI, it's kinda bad practice since observedGeneration should be per condition.

Reply:
I've added a note about automatic evaluation of observedGeneration.